// eris-giveaways/src/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BonusEntry = exports.GiveawayEditOptions = exports.GiveawayStartOptions = exports.GiveawayData = exports.GiveawayRerollOptions = exports.GiveawayManagerOptions = exports.PauseOptions = exports.LastChanceOptions = exports.InteractionOptions = exports.GiveawayMessages = exports.Endpoints = void 0;
exports.Endpoints = {
    MESSAGE_URL: (guildID, channelID, messageID) => `https://discord.com/channels/${guildID}/${channelID}/${messageID}`,
};
exports.GiveawayMessages = {
    giveaway: "🎉🎉 **GIVEAWAY** 🎉🎉",
    giveawayEnded: "🎉🎉 **GIVEAWAY ENDED** 🎉🎉",
    inviteToParticipate: "React with 🎉 to participate!",
    drawing: "Drawing: **{duration}**",
    winMessage: "Congratulations, {winners}! You won **{this.prize}**! \n {this.messageURL}",
    dropMessage: "Be the first to react 🎉 !",
    embedFooter: "Powered by givies-framework",
    noWinner: "Giveaway cancelled, no valid participations.",
    winners: "Winner(s):",
    endedAt: "Ended at",
    hostedBy: "Hosted by: {this.hostedBy}",
};
exports.InteractionOptions = {
    enabled: false,
    ephemeral: false
};
exports.LastChanceOptions = {
    enabled: true,
    content: "⚠️ **LAST CHANCE TO ENTER** ⚠️",
    threshold: 10000,
    embedColor: 0xff0000,
};
exports.PauseOptions = {
    isPaused: false,
    content: "⚠️ **THIS GIVEAWAY IS PAUSED** ⚠️",
    unPauseAfter: null,
    embedColor: 0xffff00,
    durationAfterPause: null,
    infiniteDurationText: "`NEVER`"
};
exports.GiveawayManagerOptions = {
    storage: "./giveaways.json",
    forceUpdateEvery: null,
    endedGiveawaysLifetime: null,
    default: {
        botsCanWin: false,
        exemptPermissions: [],
        exemptMembers: () => false,
        embedColor: 0x7289da,
        embedColorEnd: 0x23272a,
        reaction: "🎉",
        lastChance: {
            enabled: true,
            content: "⚠️ **LAST CHANCE TO ENTER** ⚠️",
            threshold: 10000,
            embedColor: 0xff0000,
        },
    },
};
exports.GiveawayRerollOptions = {
    winnerCount: 1,
    messages: {
        congrat: "🎉 **New winner(s):** {winners}! Congratulations, you won **{this.prize}**! \n {this.messageURL}",
        error: "No valid participations, no winners can be rerolled!",
    },
    interactionOptions: {
        enabled: false,
        ephemeral: false
    }
};
exports.GiveawayData = {
    startAt: 0,
    botsCanWin: false,
    endAt: 0,
    winnerCount: 0,
    messages: undefined,
    prize: "",
    channelID: "",
    guildID: "",
    ended: false
};
exports.GiveawayStartOptions = {
    prize: "",
    winnerCount: 0
};
exports.GiveawayEditOptions = {};
exports.BonusEntry = [];
